jenkins/PerfCI_Grafana_Deployment/Jenkinsfile

```diff
@@ -69,74 +69,91 @@ pipeline {
+                                git config --global user.name "${USER_NAME}"
+                                git config pull.rebase false  # merge
                             '''
-                            sh 'git add benchmark_runner/grafana/perf/jsonnet/main.libsonnet'
-                        }
 
-                        // Check if there are changes to commit
-                        def hasChanges = sh(script: 'git diff --exit-code', returnStatus: true) != 0
-
-                        if (hasChanges) {
-                            // Commit, pull, and push
-                            sh '''
-                                git commit -m "Update Grafana files"
-                                git pull https://${GIT_TOKEN}@${GIT_REPOSITORY} main
-                                git push https://${GIT_TOKEN}@${GIT_REPOSITORY} main
-                            '''
-                        } else {
-                            echo 'No changes to commit.'
+                            // Update dashboard.json
+                            if (CHANGES_DASHBOARD.toBoolean()) {
+                                sh '''
+                                    cp -p benchmark_runner/grafana/perf/dashboard.json benchmark_runner/grafana/perf/dashboard_backup.json;
+                                    PYTHONPATH=. sudo python3 benchmark_runner/grafana/update_grafana_dashboard.py --grafana_url="${GRAFANA_URL}" --grafana_api_key="${GRAFANA_API_KEY}" --grafana_json_path="${GRAFANA_JSON_PATH}" --grafana_folder_name="${GRAFANA_FOLDER_NAME}";
+                                    cp -p benchmark_runner/grafana/perf/dashboard_backup.json benchmark_runner/grafana/perf/dashboard.json;
+                                '''
+                                sh 'git add benchmark_runner/grafana/perf/dashboard.json'
+                            }
+
+                            // Update main.libsonnet
+                            if (CHANGES_MAIN_LIBSONNET.toBoolean()) {
+                                sh '''
+                                    cp -p benchmark_runner/grafana/perf/jsonnet/main.libsonnet benchmark_runner/grafana/perf/jsonnet/main.libsonnet_backup
+                                    # Add code to update main.libsonnet here
+                                '''
+                                sh 'git add benchmark_runner/grafana/perf/jsonnet/main.libsonnet'
+                            }
+
+                            // Check if there are changes to commit
+                            if (CHANGES_MAIN_LIBSONNET.toBoolean() || CHANGES_DASHBOARD.toBoolean()) {
+                                // Commit, pull, and push
+                                sh '''
+                                    git commit -m "Update Grafana files"
+                                    git pull https://${GIT_TOKEN}@${GIT_REPOSITORY} main
+                                    git push https://${GIT_TOKEN}@${GIT_REPOSITORY} main
```

Comment:
Need to ensure nothing goes wrong with any of these commands

Reply:
It is working properly

Reply:
What happens if it fails (e. g. because you can't connect to the git server, possibly just one command failing)?

Reply:
We will do it in the next round - run twice a week, also the stage will fail because all the code in try and catch

Reply:
It won't fail unless the git push is what fails.  An easy way to do this would be

```
git commit -m "Update Grafana files" && git pull https://${GIT_TOKEN}@${GIT_REPOSITORY} main && git push https://${GIT_TOKEN}@${GIT_REPOSITORY} main"
```

That ensures that if any command fails that the script immediately fails rather than continuing.

Reply:
try & catch should wrap all the sh script, not only the first command